### PR TITLE
Don't cache ValueSource's for producing trees for DocumentStates that don't 'Support Syntax Tree'.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This ends up creating an unending chain of ValueSources to lazily realize a tree for a
document that will never end up producing a tree.  This manifests as a very slow memory
leak.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
This is obviously a very TypeScript specific issue.  In the C#/VB world you guys obviously have features that ask for SyntaxTrees, so you very quickly collapse thse chains.  In the TS world, there is nothing that ever gets a SyntaxTree, so the _TreeSource is never 'pulled' on, and so it always just holds onto it's lambda, which holds onto an old tree source, which holds onto a lambda, which holds onto an old tree source, etc. etc.

The fix here is for us to not even create these lazy chains for documents that won't ever retrieve SyntaxTrees.

**Comment by CyrusNajmabadi** [comment]
Paging @heejaechang  and @mattwar .

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs [hunk 1]
```diff
@@ -342,14 +363,19 @@ public new DocumentState UpdateText(TextLoader loader, PreservationMode mode)
...
-                this.Id.ProjectId,
-                GetSyntaxTreeFilePath(this.info),
-                _options,
-                _languageServices,
-                this.solutionServices,
-                mode);
+            // Only create the ValueSource for creating the SyntaxTree if this is a Document that
+            // supports SyntaxTrees.  There's no point in creating the async lazy and holding onto
+            // this data otherwise.
+            var newTreeSource = !this.SupportsSyntaxTree
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs]: small thing but can these two "?:" use same order of condition, either this.SupportSyntax .. or !this.SupportSyntax...

**Comment by heejaechang** [comment]
:+1:

**Comment by mattwar** [comment]
:+1:

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs]: Sure!  I can fix that up.

**Pull request merged**